runtime-extension-model/src/main/java/org/mule/runtime/core/internal/extension/ForEachCollectionTypeResolver.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2023 Salesforce, Inc. All rights reserved.
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.core.internal.extension;
+
+import static org.mule.runtime.api.metadata.resolving.FailureCode.UNKNOWN;
+import static org.mule.runtime.core.api.extension.provider.MuleExtensionModelProvider.ANY_TYPE;
+
+import static java.lang.String.format;
+
+import org.mule.metadata.api.model.ArrayType;
+import org.mule.metadata.api.model.MetadataType;
+import org.mule.metadata.message.api.el.TypeBindings;
+import org.mule.runtime.api.connection.ConnectionException;
+import org.mule.runtime.api.metadata.ExpressionLanguageMetadataService;
+import org.mule.runtime.api.metadata.MetadataContext;
+import org.mule.runtime.api.metadata.MetadataResolvingException;
+import org.mule.runtime.api.metadata.resolving.InputTypeResolver;
+import org.mule.runtime.privileged.metadata.InternalMetadataContext;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * {@link InputTypeResolver} implementation used for the "collection" parameter of the ForEach and ParallelForEach scopes.
+ *
+ * @since 4.8.0
+ */
+public class ForEachCollectionTypeResolver implements InputTypeResolver<String> {
+
+  @Override
+  public String getCategoryName() {
+    return "FOREACH";
+  }
+
+  @Override
+  public String getResolverName() {
+    return "FOREACH_INPUT_COLLECTION";
+  }
+
+  @Override
+  public MetadataType getInputMetadata(MetadataContext context, String key)
+      throws MetadataResolvingException, ConnectionException {
+    if (context instanceof InternalMetadataContext) {
```

Comment:
it should fail if not true. Otherwise you're hiding a bug

Reply:
We will only have this in the context of DietTooling right now. So, I'm not sure failing is a good idea, unless we are sure MTS is not used for a ForEach.

Reply:
I forgot to mention it yesterday, but I already changed this one. I made sure all our current `MetadataContext` implementations (which were not that many) implement `InternalMetadataContext` too.